WMass/python/postprocessing/examples/lepSFProducer.py

```diff
@@ -115,11 +115,162 @@ def analyze(self, event):
+
+    def endJob(self):
+        pass
+    def beginFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
+        self.out = wrappedOutputTree
+        self.out.branch("LepGood_IDeffSF",      "F", lenVar="nLepGood")
+        self.out.branch("LepGood_ISOeffSF",     "F", lenVar="nLepGood")
+        self.out.branch("LepGood_trgSF",        "F", lenVar="nLepGood")
+        self.out.branch("LepGood_IDeffSF_err",  "F", lenVar="nLepGood")
+        self.out.branch("LepGood_ISOeffSF_err", "F", lenVar="nLepGood")
+        self.out.branch("LepGood_trgSF_err",    "F", lenVar="nLepGood")
+    def endFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
+        pass
+    def analyze(self, event):
+        """process event, return True (go to next module) or False (fail, go to next event)"""
+        leps = Collection(event, "LepGood")
+        sf_id  = []
+        sf_iso = []
+        sf_trg = []
+        sf_id_err  = []
+        sf_iso_err = []
+        sf_trg_err = []
+        for l in leps:
+            if event.isData:
+                sf_trg.append(1.)
+                sf_id.append(1.)
+                sf_iso.append(1.)
+                sf_trg_err.append(0.)
+                sf_id_err.append(0.)
+                sf_iso_err.append(0.)
+            else:
+                if abs(l.pdgId)==11:
+                    sf_trg.append(float(self.trgSF_manager_el.getSF(l.pt,l.eta)))                    
+                    sf_id.append(1.)
+                    sf_iso.append(1.)
+                    sf_trg_err.append(float(self.trgSF_manager_el.getSF_err(l.pt,l.eta)))                    
+                    sf_id_err.append(1.)
+                    sf_iso_err.append(1.)
+                else:
+                    sf_trg.append(    float(self.trgSF_manager_mu.getSF(l.pt,l.eta)))
+                    sf_id.append(     float(self.idSF_manager_mu.getSF(l.pt,l.eta)))
+                    sf_iso.append(    float(self.isoSF_manager_mu.getSF(l.pt,l.eta)))
+                    sf_trg_err.append(float(self.trgSF_manager_mu.getSF_err(l.pt,l.eta)))
+                    sf_id_err.append( float(self.idSF_manager_mu.getSF_err(l.pt,l.eta)))
+                    sf_iso_err.append(float(self.isoSF_manager_mu.getSF_err(l.pt,l.eta)))
+        self.out.fillBranch("LepGood_IDeffSF",  sf_id)
+        self.out.fillBranch("LepGood_ISOeffSF", sf_iso)
+        self.out.fillBranch("LepGood_trgSF",    sf_trg)
+        self.out.fillBranch("LepGood_IDeffSF_err",  sf_id_err)
+        self.out.fillBranch("LepGood_ISOeffSF_err", sf_iso_err)
+        self.out.fillBranch("LepGood_trgSF_err",    sf_trg_err)
+        return True
+
+#########################################################
+
+
 # define modules using the syntax 'name = lambda : constructor' to avoid having them loaded when not needed
 
 lepSF = lambda : lepSFProducer( "TightWP_2016", "CutBasedTight_2016")
 trgSF = lambda : lepTrgSFProducer()
+lep2016SF = lambda : lep2016SFProducer()
```

Comment:
I think that we need to run two of them, one for the nominal SF, and one for the alternative SF, and then treat the different reweighting as the systematic uncertainty. 
Here the single histogram uncertainty is the stat error of the fit?

Reply:
The uncertainty is currently just the uncertainty from the points, which means a constant uncertainty in each range defined by each bin width along the x axis. Especially for the ID, where points are almost randomly distributed (but sometimes with big uncertainties) it seemed the safest definition

There cannot be any sensible definition of the uncertainty from the fit, because the fit gives you uncertainty on the parameters, and they do not account just for offset variations, but also for change of curvature and similar. Once we have the alternative scale factors, either we define the uncertainty as the difference of the two sets of points or the difference of the two fits (if the points are not weird, the two solutions should be equivalent to each other)

Reply:
so summarizing the answer: YES

Reply:
having two scale factors, so that we can reweight the sample twice and have it as shape uncertainty. But let's leave it aside and have just one to see the level of the closure.

Reply:
ok, if we want 2 we could just use the variable of the uncertainty to store the second scale factor (no need to store uncertainty as well, if that is defined as the scale factor difference).
Otherwise we would end up with 12 variables (6 scale factors and their uncertainies)

Reply:
Yes, let's see how big they are. I would take the largest between the alternative and the statistical error of the (di-lep mass) fit. Which I think will be the alternative, since the stat error should be very small for pt ~40 GeV.